cupy/manipulation/basic.py

```diff
@@ -69,3 +69,61 @@ def _can_memcpy(dst, src):
     f_contiguous = dst.flags.f_contiguous and src.flags.f_contiguous
     return (c_contiguous or f_contiguous) and dst.dtype == src.dtype and \
         dst.size == src.size
+
+
+_putmask_kernel = core._kernel.ElementwiseKernel(
+    'Q mask, raw T values, uint64 len_vals', 'T out',
+    '''
+    if (mask) out = values[i % len_vals];
+    ''',
+    'putmask_kernel'
+)
+
+
+def putmask(a, mask, values):
+    """
+    Changes elements of an array inplace, based on conditional mask and
+    input values.
+
+    Sets ``a.flat[n] = values[n]`` for each n where ``mask.flat[n]==True``.
+    If `values` is not the same size as `a` and `mask` then it will repeat.
+
+    Args
+        a (cupy.ndarray): Target array.
+        mask (cupy.ndarray): Boolean mask array. It has to be
+            the same shape as `a`.
+        values (cupy.ndarray): Values to put into `a` where `mask` is True.
```

Comment:
For a frequent case, please support scalar `values`.
```suggestion
        values (cupy.ndarray or scalar): Values to put into `a` where `mask` is True.
```

Reply:
Okay, will do that. But in case you haven't noticed my latest comments above, can you please respond to them? @asi1024 
So that I can commit both the changes together. See https://github.com/cupy/cupy/pull/3261#issuecomment-610927701 and https://github.com/cupy/cupy/pull/3261#issuecomment-610931251

Reply:
How about using `numpy.can_cast`?

Reply:
I have already used that in testing. But the tests are failing for even safe casting from numpy.int16 to numpy.int64. It says `casting is not supported yet`.

Reply:
So, how about using unsafe casting, checking types with `numpy.can_cast` and raising errors if needed?

Reply:
BTW, I think you can implement it without this typecast. Please avoid it for the performance.

Reply:
>BTW, I think you can implement it without this typecast. Please avoid it for the performance.

Okay, so I am removing that casting and the test with different dtypes.

Reply:
No, don't remove the tests.

Reply:
But they will fail on the kernel if the dtypes of `a` and `values` are different.

Reply:
Please fix to pass tests without typecasting. If you cannot, it's OK to just leave a TODO.

Reply:
Can you take a look now?

Reply:
I mean, please keep the compatibility to NumPy without typecasting.

Reply:
okay, then would casting the data inside the Kernel perform better?

Reply:
>okay, then would casting the data inside the Kernel perform better? I mean I have seen this somewhere in the library.

I did this. Can you please take a look and tell me whether it is a good way or not?